File: core/blas/colwise_mul.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and contiditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "colwise_mul.h"
#include "blas_common.h"
#include "blasplan/include/blasplan/BlasColwiseMulPlan.h"

using namespace AsdSip;
using namespace Mki;

namespace AsdSip {
struct ColwiseMuParam {
    int64_t m;
    int64_t n;
    aclTensor *mat;
    aclTensor *vec;
    aclTensor *result;
};

AspbStatus ColwiseMulDtypeCheck(struct ColwiseMuParam parm)
{
    auto ret = ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE;
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.mat, aclDataType::ACL_COMPLEX64, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.vec, aclDataType::ACL_COMPLEX64, ret);
    SIP_OP_CHECK_DTYPE_NOT_MATCH(parm.result, aclDataType::ACL_COMPLEX64, ret);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus ColwiseMulShapeCheck(struct ColwiseMuParam parm)
{
    int64_t m = parm.m;
    int64_t n = parm.n;
    ASDSIP_ECHECK(m > 0 && n > 0 && m <= UINT32_MAX && n <= UINT32_MAX,
                  "blas ColwiseMul get m <= 0 || n <= 0 or m or n > 2^32.", ErrorType::ACL_ERROR_INVALID_PARAM);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.mat, m * n, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.vec, m, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    SIP_OP_CHECK_NUM_NOT_MATCH(parm.result, m * n, ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);
    return ErrorType::ACL_SUCCESS;
}

AspbStatus asdBlasColwiseMul(asdBlasHandle handle, const int64_t m, const int64_t n, aclTensor *mat,
                             aclTensor *vec, aclTensor *result)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(BlasPlanCache::doesPlanExist(handle),
                  "blas ColwiseMul get cached plan failed.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

    ColwiseMuParam implParam = {m, n, mat, vec, result};
    ASDSIP_CHECK(ColwiseMulDtypeCheck(implParam) == ErrorType::ACL_SUCCESS,
                 "blas asdBlasColwiseMul dtype check failed.", return ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    ASDSIP_CHECK(ColwiseMulShapeCheck(implParam) == ErrorType::ACL_SUCCESS,
                 "blas asdBlasColwiseMul shape check failed.", return ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH);

    try {
        AsdSip::BlasColwiseMulPlan &plan = dynamic_cast<AsdSip::BlasColwiseMulPlan &>(BlasPlanCache::getPlan(handle));
        ASDSIP_ECHECK(plan.IsInitialized(), "ColwiseMul plan init Error!.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

        Status status;
        OpDesc opDesc;
        opDesc.opName = "ColwiseMulOperation";
        AsdSip::OpParam::ColwiseMul param;
        param.m = m;
        param.n = n;
        opDesc.specificParam = param;
        ASDSIP_LOG(DEBUG) << "OpDesc: " << opDesc.opName << "; OpDesc info: " << param.ToString();

        SVector<aclTensor *> colwiseMulInTensors{mat, vec, plan.augAclTensor};
        SVector<aclTensor *> colwiseMulOutTensors{result};

        status = RunAsdOpsV2(plan.GetStream(), opDesc, colwiseMulInTensors, colwiseMulOutTensors, plan.GetWorkspace());
        ASDSIP_ECHECK(status.Ok(), status.Message(), ErrorType::ACL_ERROR_INTERNAL_ERROR);

        ASDSIP_LOG(INFO) << "Execute asdBlasColwiseMul success.";
        return ErrorType::ACL_SUCCESS;
    } catch (std::bad_cast &e) {
        // Handle the BlasColwiseMul exception
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "ColwiseMul Error: " << e.what();
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
}

AspbStatus asdBlasMakeColwiseMulPlan(asdBlasHandle handle)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(handle != nullptr, "blas MakeColwiseMulPlan Fail.", ErrorType::ACL_ERROR_INTERNAL_ERROR);
    AsdSip::BlasColwiseMulPlan *plan = nullptr;
    try {
        plan = new AsdSip::BlasColwiseMulPlan();
        BlasPlanCache::MakePlan(handle, plan);
    } catch (const std::exception& e) {
        if (plan != nullptr) {
            delete plan;
        }
        delete static_cast<int*>(handle);
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Make ColwiseMul Plan failed: " << e.what();
        throw std::runtime_error("Make ColwiseMul Plan failed.");
    }
    if (plan->CreateTensor() != ErrorType::ACL_SUCCESS) {
        plan->MarkFailed();
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Fail to create blas asdBlasMakeColwiseMulPlan mask tensor.";
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
    plan->MarkInitialized();
    return ErrorType::ACL_SUCCESS;
}
}

File: sip_pta/test/blas/colwise_mul.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Copyright (c) 2026 Huawei Technologies Co., Ltd.
This program is free software, you can redistribute it and/or modify it under the terms and conditions of
CANN Open Software License Agreement Version 2.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
"""

import logging
import os
import sys
from dataclasses import dataclass
from typing import Tuple

import torch
import torch_npu
import torch_sip

# 配置全局日志 (G.LOG.02)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[logging.StreamHandler(sys.stdout)]
)
logger = logging.getLogger("torch_sip_colwise_mul")


@dataclass(frozen=True)
class ColwiseMulConfig:
    """参数对象：封装矩阵按列乘法的维度配置"""
    m: int
    n: int
    name: str = "Case"


class ColwiseMulTester:
    """矩阵按列与向量逐元素相乘测试类"""

    def __init__(self, device: str = "npu:0"):
        self.device = device
        blocking_env = os.getenv("ASCEND_LAUNCH_BLOCKING", "0")
        mode_str = "异步 (Async)" if blocking_env == "0" else "同步 (Blocking)"
        logger.info("当前 NPU 运行模式: %s (ASCEND_LAUNCH_BLOCKING=%s)", mode_str, blocking_env)

    @staticmethod
    def get_complex_tensor(shape: Tuple[int, ...], device: str) -> torch.Tensor:
        """
        静态方法：安全生成复数 Tensor。
        """
        real = torch.randn(shape, dtype=torch.float32)
        imag = torch.randn(shape, dtype=torch.float32)
        return torch.complex(real, imag).to(device)

    def run_case(self, cfg: ColwiseMulConfig) -> bool:
        """
        执行单次测试用例。
        G.ERR.01: 最小化 try 块，仅包裹算子核心逻辑。
        """
        # 1. 准备数据
        mat = self.get_complex_tensor((cfg.m, cfg.n), self.device)
        vec = self.get_complex_tensor((cfg.m,), self.device)

        # 备份用于 CPU 参考计算
        mat_cpu = mat.cpu()
        vec_cpu = vec.cpu()

        # 确保内存连续
        mat_in = mat.contiguous()
        vec_in = vec.contiguous()

        # 2. 核心调用 (G.ERR.01: 仅包裹 NPU 算子执行)
        try:
            # 根据注册方式调用算子
            out_npu = torch_sip.asd_blas_colwise_mul(mat_in, vec_in)
        except Exception as exc:
            logger.error("[%s] 算子执行崩溃: %s", cfg.name, exc)
            return False

        # 3. CPU 参考计算 (移出 try 块以明确异常源)
        # mat (m, n) * vec (m, 1) 利用广播机制实现按列相乘
        ref = mat_cpu * vec_cpu.unsqueeze(1)

        # 4. 比对验证
        out_cpu = out_npu.cpu()
        is_close = torch.allclose(out_cpu, ref, rtol=1e-4, atol=1e-4)

        status = "PASS" if is_close else "FAIL"
        logger.info("[%s] %-20s | M=%d, N=%d", status, cfg.name, cfg.m, cfg.n)

        if not is_close:
            max_diff = torch.abs(out_cpu - ref).max()
            logger.error("      Max Diff: %.6e", max_diff)

        return is_close


def main():
    """主测试套件入口"""
    tester = ColwiseMulTester()
    logger.info("开始 ColwiseMul 专项测试 ...\n")

    # 使用参数对象定义测试套件
    test_suites = [
        ColwiseMulConfig(8, 4, "Small_Matrix"),
        ColwiseMulConfig(16, 16, "Square_Matrix"),
        ColwiseMulConfig(128, 64, "Large_Rectangle"),
        ColwiseMulConfig(256, 1, "Single_Column"),
        ColwiseMulConfig(1, 256, "Single_Row"),
    ]

    all_passed = True
    for config in test_suites:
        if not tester.run_case(config):
            all_passed = False

    logger.info("-" * 60)
    if all_passed:
        logger.info("测试结论: ✅ 全部通过")
    else:
        logger.error("测试结论: ❌ 存在失败项")

    return 0 if all_passed else 1


if __name__ == "__main__":
    # 整个脚本唯一的退出出口
    sys.exit(main())